utils/densify.go

```diff
@@ -1,110 +1,195 @@
 package utils
 
 import (
-	"math"
+	"errors"
+	"fmt"
 
 	geod "github.com/starboard-nz/go-geodesy"
 	"github.com/starboard-nz/orb"
 	"github.com/starboard-nz/units"
 )
 
+var (
+	ErrInvalidTolerance = errors.New("invalid value for tolerance - must be positive")
+	ErrToleranceTooLow  = errors.New("tolerance too low")
+	ErrInternalError    = errors.New("internal error")
+	ErrInvalidGeometry  = errors.New("invalid geometry")
+)
+
 // NOTE - these densify functions will only work as expected if passing geometries in the normalised (-180 to 180) range.
 
 // DensifyMultiPolygon inserts points into the multipolygon using the given Model, until the maximum distance between
-// planar geometry and the given model is less than the tolerance.
-func DensifyMultiPolygon(mp orb.MultiPolygon, tolerance units.Distance, earthModel func(geod.LatLon, ...interface{}) geod.Model) orb.MultiPolygon {
-	var dmp orb.MultiPolygon
+// model and the reference model is less than the tolerance, where model defines the shape of the lines between points
+// (e.g. great circle arc or rhumb line).
+func DensifyMultiPolygon(mp orb.MultiPolygon, model, refModel geod.EarthModel, tolerance units.Distance) (orb.MultiPolygon, error) {
+	var (
+		dmp orb.MultiPolygon
+		err error
+	)
 
 	for _, polygon := range mp {
-		dmp = append(dmp, DensifyPolygon(polygon, tolerance, earthModel))
+		dp, err2 := DensifyPolygon(polygon, model, refModel, tolerance)
+		if err2 != nil {
+			if !errors.Is(err2, ErrToleranceTooLow) {
+				return nil, err2
+			}
+
+			err = err2
+		}
+
+		dmp = append(dmp, dp)
 	}
 
-	return dmp
+	return dmp, err
 }
 
 // DensifyPolygon inserts points into the polygon using the given Model, until the maximum distance between
 // planar geometry and the given model is less than the tolerance.
-func DensifyPolygon(poly orb.Polygon, tolerance units.Distance, earthModel func(geod.LatLon, ...interface{}) geod.Model) orb.Polygon {
-	var dp orb.Polygon
+func DensifyPolygon(poly orb.Polygon, model, refModel geod.EarthModel, tolerance units.Distance) (orb.Polygon, error) {
+	var (
+		dp  orb.Polygon
+		err error
+	)
 
 	for _, ring := range poly {
-		dp = append(dp, DensifyRing(ring, tolerance, earthModel))
+		dr, err2 := DensifyRing(ring, model, refModel, tolerance)
+		if err2 != nil {
+			if !errors.Is(err2, ErrToleranceTooLow) {
+				return nil, err2
+			}
+
+			err = err2
+		}
+
+		dp = append(dp, dr)
 	}
 
-	return dp
+	return dp, err
 }
 
 // DensifyRing inserts points into the ring using the given Model, until the maximum distance between
 // planar geometry and the given model is less than the tolerance.
-func DensifyRing(ring orb.Ring, tolerance units.Distance, earthModel func(geod.LatLon, ...interface{}) geod.Model) orb.Ring {
+func DensifyRing(ring orb.Ring, model, refModel geod.EarthModel, tolerance units.Distance) (orb.Ring, error) {
 	if len(ring) < 2 {
```

Comment:
Shouldn't this be `< 4`? I mean technically we should ask for `< 5` but I don't want to enforce rings returning to the same initial point here.